### DISCUSSION TITLE
How to display an SKBitmap (or any direct pixel buffer representation) in MAUI?

### DISCUSSION THREAD

**Discussion by koszeggy** [dotnet/maui#8490] [Q&A]
In other XAML-based environments (such as WPF, UWP, Avalonia, Uno) I could just use a `WriteableBitmap` type, which combines two good features: allows direct access to the pixel buffer and also can be used as an image source for the regular `Image` view.

In contrast, MAUI appears to go on the Xamarin way and uses Skia's solutions (at least judging by the numerous [NuGet packages](https://www.nuget.org/packages?q=skia+maui)). And though displaying an `SKBitmap` was [possible](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/bitmaps/displaying) with using a specialized view (provided by mono/SkiaSharp), I was [sweating blood](https://github.com/mono/SkiaSharp/issues/2139) while struggling with the same task in MAUI, without any success.

I was hesitating whether to create a new feature request instead (add `WriteableBitmap`) because I'm not even sure whether I just misused something or are these real bugs in the preview version that have to work in the final release. I even started to create a poll to see whether others would prefer a `WriteableBitmap` type in MAUI, but in the end I reverted to a simple question to see first how to do it correctly.

> 📝 **Note:** See my detailed attempts with code samples and the attached project in [this](https://github.com/mono/SkiaSharp/issues/2139) bug report in the SkiaSharp repo. Here I just summarize the TL;DR version of my attempts to display an `SKBitmap`:
1. The Xamarin way: Using `SKCanvasView` from SkiaSharp.Views.Maui.Controls and handling its `OnPaintSurface`
2. The most appealing way: setting `Image.Source` to an `SKBitmapImageSource` that wraps by bitmap
3. Using `GraphicsView` from Microsoft.Maui.Controls and setting its `Drawable` to a `SkiaImage` that wraps my bitmap
4. Using `GraphicsView` with a custom `IDrawable` implementation as per Microsoft's [guideline](https://docs.microsoft.com/en-us/dotnet/maui/user-interface/controls/graphicsview) and draw the bitmap from there

**Disclaimer:** I know that encoding as .png + decoding again by `PlatformImage.FromStream` would "work" but unfortunately that is an unacceptable solution for me. It is a no-go in situations where I would normally use a `WriteableBitmap`.

**Comment by rokenbuzz** [comment]
@koszeggy I found your discussion here after struggling in nearly the same 4 ways you describe. All roads lead me to dead ends where I'm left scratching my head. Seemingly incompatible pieces and silent failures. Every reasonable chain of methods has a break in it. It's maddening.
Ultimately I want to be able to load a jpg into a bitmap, manipulate the pixels and display it, in a C# app that can be built for windows, android, and ios.
It has been a month since your post here. Have you had any breakthroughs you can share?

**Reply by koszeggy** [reply]
@rokenbuzz: There is a possible solution [here](https://github.com/mono/SkiaSharp/issues/2139#issuecomment-1176088865), though I had no time to try it so far.

**Reply by rokenbuzz** [reply]
Thank you. I just tried that and I'm having success on Windows and Android (emulator). I don't have iOS handy at the moment.
